protocols/autonatv2/src/client/behaviour.rs

```diff
@@ -0,0 +1,350 @@
+use std::{
+    collections::{HashMap, HashSet, VecDeque},
+    task::{Context, Poll},
+    time::Duration,
+};
+
+use either::Either;
+use futures::FutureExt;
+use futures_timer::Delay;
+use libp2p_core::{multiaddr::Protocol, transport::PortUse, Endpoint, Multiaddr};
+use libp2p_identity::PeerId;
+use libp2p_swarm::{
+    behaviour::{ConnectionEstablished, ExternalAddrConfirmed},
+    ConnectionClosed, ConnectionDenied, ConnectionHandler, ConnectionId, DialFailure, FromSwarm,
+    NetworkBehaviour, NewExternalAddrCandidate, NotifyHandler, ToSwarm,
+};
+use rand::{seq::SliceRandom, Rng};
+use rand_core::{OsRng, RngCore};
+
+use crate::{global_only::IpExt, request_response::DialRequest};
+
+use super::handler::{
+    dial_back,
+    dial_request::{self, StatusUpdate},
+    TestEnd,
+};
+
+#[derive(Debug, Clone, Copy)]
+pub struct Config {
+    pub(crate) test_server_count: usize,
+    pub(crate) max_addrs_count: usize,
+    pub(crate) recheck_interval: Duration,
+}
+
+impl Config {
+    pub fn with_test_server_count(self, test_server_count: usize) -> Self {
+        Self {
+            test_server_count,
+            ..self
+        }
+    }
+
+    pub fn with_max_addrs_count(self, max_addrs_count: usize) -> Self {
+        Self {
+            max_addrs_count,
+            ..self
+        }
+    }
+
+    pub fn with_recheck_interval(self, recheck_interval: Duration) -> Self {
+        Self {
+            recheck_interval,
+            ..self
+        }
+    }
+}
+
+impl Default for Config {
+    fn default() -> Self {
+        Self {
+            test_server_count: 3,
+            max_addrs_count: 10,
+            recheck_interval: Duration::from_secs(5),
+        }
+    }
+}
+pub struct Behaviour<R = OsRng>
+where
+    R: RngCore + 'static,
+{
+    local_peers: HashSet<ConnectionId>,
+    pending_nonces: HashSet<u64>,
+    known_servers: Vec<PeerId>,
+    rng: R,
+    config: Config,
+    pending_events: VecDeque<
+        ToSwarm<
+            <Self as NetworkBehaviour>::ToSwarm,
+            <<Self as NetworkBehaviour>::ConnectionHandler as ConnectionHandler>::FromBehaviour,
+        >,
+    >,
+    address_candidates: HashMap<Multiaddr, usize>,
+    already_tested: HashSet<Multiaddr>,
+    peers_to_handlers: HashMap<PeerId, ConnectionId>,
+    next_tick: Delay,
+}
+
+impl<R> NetworkBehaviour for Behaviour<R>
+where
+    R: RngCore + 'static,
+{
+    type ConnectionHandler = Either<dial_request::Handler, dial_back::Handler>;
+
+    type ToSwarm = StatusUpdate;
```

Comment:
We typically call the types that we return here `Event`.

Reply:
Renamed it